programs/bpf_loader/src/lib.rs

```diff
@@ -390,13 +390,27 @@ fn process_loader_upgradeable_instruction(
                 return Err(InstructionError::InvalidArgument);
             }
 
-            let instruction = system_instruction::create_account(
+            if invoke_context.is_feature_active(&fix_write_privs::id()) {
+                // Drain the Buffer account to payer before paying for programdata account
+                payer
+                    .try_account_ref_mut()?
+                    .checked_add_lamports(buffer.lamports()?)?;
+                buffer.try_account_ref_mut()?.set_lamports(0);
+            }
+
+            let mut instruction = system_instruction::create_account(
                 payer.unsigned_key(),
                 programdata.unsigned_key(),
                 1.max(rent.minimum_balance(programdata_len)),
                 programdata_len as u64,
                 program_id,
             );
+
+            // pass an extra account to avoid the overly strict UnbalancedInstruction error
```

Comment:
FYI @jackcmay I'm bumping into this issue now: https://github.com/solana-labs/solana/issues/9711